### PR TITLE
fix(page.evaluate): better function serialization

### LINKED ISSUE #1665: async function inside evaluate fails
#### Code
```
await page.evaluate(async () => {
    console.log('1')
})
```
#### Error
```
(node:66796) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Evaluation failed: ReferenceError: fn is not defined
    at <anonymous>:1:26
```

### PR DESCRIPTION
**Pull request by aslushnikov**
References #1665

### CODE REVIEW

File: lib/ExecutionContext.js [hunk 1]
```diff
@@ -94,8 +94,23 @@ class ExecutionContext {
     if (typeof pageFunction !== 'function')
       throw new Error('The following is not a function: ' + pageFunction);
 
+    let functionText = pageFunction.toString();
+    try {
+      eval(functionText);
+    } catch (e1) {
+      // This means we have a function shorthand. Try another
+      // time prefixing 'function '.
+      functionText = 'function ' + functionText;
```
Comment (JoelEinbinder) [inline lib/ExecutionContext.js]: What about async static generators?

File: lib/ExecutionContext.js [hunk 2]
```diff
@@ -94,8 +94,23 @@ class ExecutionContext {
     if (typeof pageFunction !== 'function')
       throw new Error('The following is not a function: ' + pageFunction);
 
+    let functionText = pageFunction.toString();
+    try {
+      eval(functionText);
```
Comment (JoelEinbinder) [inline lib/ExecutionContext.js]: I'd be worried about unintended side effects here. Consider: `function functionText(){return 5}`.

**Comment by JoelEinbinder** [comment]
Can we also add a note about `toString`ing the pageFunctions? This has come up many times before.

File: lib/ExecutionContext.js (hunk 1 above)
Comment (aslushnikov) [reply lib/ExecutionContext.js]: async done; the rest is too exotic for now!

File: lib/ExecutionContext.js (hunk 2 above)
Comment (aslushnikov) [reply lib/ExecutionContext.js]: good catch; switched to `new Function` instead

File: lib/ExecutionContext.js [hunk 3]
```diff
@@ -94,8 +94,26 @@ class ExecutionContext {
     if (typeof pageFunction !== 'function')
       throw new Error('The following is not a function: ' + pageFunction);
 
+    let functionText = pageFunction.toString();
+    try {
+      new Function('(' + functionText + ')');
+    } catch (e1) {
+      // This means we have a function shorthand. Try another
```
Comment (JoelEinbinder) [inline lib/ExecutionContext.js]:
```suggestion
      // This means we might have a function shorthand. Try another
```

**Review by JoelEinbinder — APPROVED** [review]

**Pull request merged**